core/src/main/scala/org/broadinstitute/dsde/rawls/user/UserService.scala

```diff
@@ -649,6 +575,150 @@ class UserService(protected val userInfo: UserInfo, val dataSource: SlickDataSou
+
+          throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "You may not remove your own access."))
+        }
+
+        DBIO.successful(rawlsGroupForRole(role, managedGroup))
+      }
+    } flatMap { rawlsGroup =>
+      updateGroupMembers(rawlsGroup, addMemberList, removeMemberList)
+    }
+  }
+
+  private def rawlsGroupForRole(role: ManagedRole, managedGroup: ManagedGroup): RawlsGroup = {
+    role match {
+      case ManagedRoles.Owner => managedGroup.ownersGroup
+      case ManagedRoles.User => managedGroup.usersGroup
+    }
+  }
+
+  def overwriteManagedGroupMembers(groupRef: ManagedGroupRef, role: ManagedRole, memberList: RawlsGroupMemberList): Future[PerRequestMessage] = {
+    dataSource.inTransaction { dataAccess =>
+      withManagedGroupOwnerAccess(groupRef, RawlsUser(userInfo), dataAccess) { managedGroup =>
+        if (role == ManagedRoles.Owner &&
+          !memberList.userEmails.getOrElse(Seq.empty).contains(userInfo.userEmail) &&
+            !memberList.userSubjectIds.getOrElse(Seq.empty).contains(userInfo.userSubjectId)) {
+
+          throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.BadRequest, "You may not remove your own access."))
+        }
+
+        DBIO.successful(rawlsGroupForRole(role, managedGroup))
+      }
+    } flatMap { rawlsGroup =>
+      overwriteGroupMembers(rawlsGroup, memberList)
+    }
+  }
+
+  def deleteManagedGroup(groupRef: ManagedGroupRef) = {
+    // note that this function does not call deleteGroup (in this class) which does the desired work of deleting
+    // a rawls group and associated google group because we need to catch any FK constraint violations caused by
+    // deleting the rawls groups and if there are any rollback the db delete and not remove google groups
+    for {
+      groupEmailsToDelete <- dataSource.inTransaction { dataAccess =>
+        withManagedGroupOwnerAccess(groupRef, RawlsUser(userInfo), dataAccess) { managedGroup =>
+          DBIO.seq(
+            dataAccess.managedGroupQuery.deleteManagedGroup(groupRef),
+            dataAccess.rawlsGroupQuery.delete(managedGroup.usersGroup),
+            dataAccess.rawlsGroupQuery.delete(managedGroup.ownersGroup)
+          ).map(_ => Seq(managedGroup.usersGroup, managedGroup.ownersGroup))
+        }
+      }.recover {
+        // assume any sql exception is a FK violation, to be more specific catch MySQLIntegrityConstraintViolationException
+        // but it is good to keep the code free of mysql specific code
+        case sqle: SQLException => throw new RawlsExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, "Cannot delete group because it is in use", sqle))
+      }
+
+      _ <- Future.traverse(groupEmailsToDelete) { group =>
+        gcsDAO.deleteGoogleGroup(group).recover {
+          // log any exception but ignore
+          case t: Throwable => logger.error(s"error deleting google group $group", t)
```

Comment:
We can't do that because an existing group may have members and then it will be out of sync. This particular bit of code is a best effort cleanup. It is difficult because the database entries are already gone at this point so there is not much else to do but move on. I could include the google deletes in the transaction, but that is bad practice, plus what if 1 group is deleted and another fails... also a bad state. We can't delete the google groups first because it might be in use.

This seems like the least evil/complicated code.